### ISSUE TITLE
Feasibility of Inline Assembly in Place of DynASM

### ISSUE THREAD

**Issue by JanByrne** [LuaJIT/LuaJIT#887]
If the assembler source code for LuaJIT 2's interpreter were written using GNU inline assembler or such rather than DynASM syntax-assembler, I believe it would simplify the build process by removing the requirement for a host compiler (used to compile minilua for DynASM preprocessing and buildvm for executable file generation), which has become a problem recently when it comes to cross-compiling for different bitness, e.g MacOS and Rosetta 2. It would also make it easier to port LuaJIT since DynASM wouldn't need to be ported (as it's not used in a JIT capacity anymore).

I'm probably missing something; I hope this isn't Dunning–Kruger effect.

**Comment by MikePall** [comment]
> I'm probably missing something

Yes.

> I hope this isn't Dunning–Kruger effect.

No, it's not, because you're aware that you're missing something.

But on to your question:

"GNU inline assembler or such" is a slight understatement of the task. Apart from not working on MSVC, the assembler part of the interpreter cannot be mapped easily to a regular C program. It doesn't have function boundaries, it has non-traditional control-flow, it needs specific landing and exit pads and it has unusual unwind info. Trying to fight with the C compiler to produce the exact code required is a game you cannot win.

Of course, one could write a 'traditional' interpreter in C, but then it would be very hard to match the speed of the current LuaJIT interpreter.

The 32/64 bit structure sizing issue could be worked around in a different way, if really required.

tl;dr: Use Linux for development -- it's great. In a container, if you must.
